### PR TITLE
Implement correct corner clipping for background color

### PR DESCRIPTION
**Pull request by pyfisch**
Add one regression ref test.

See also #19649

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19675)

### CODE REVIEW

**Comment by pyfisch** [comment]
r? @emilio 

@bors-servo try

I need help to also support background images. They already have a clip from CSS as a local clip and the corner clipping should attach another clip. How is this best done?

File: tests/wpt/web-platform-tests/css/css-backgrounds/background-clip/clip-rounded-corner-ref.html [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        width: 30px;
+        height: 30px;
+        left: 120px;
+        top: 5px;
+        background-color: black;
+    }
+</style>
+<body>
+    <div id="a"></div>
+    <div id="shield"></div>
```
Comment (emilio) [inline tests/wpt/web-platform-tests/css/css-backgrounds/background-clip/clip-rounded-corner-ref.html]: Maybe the `shield` comment is useful here too?

**Review by emilio — APPROVED** [review]
LGTM with those comments addressed. Thanks a lot!

File: tests/wpt/web-platform-tests/css/css-backgrounds/background-clip/clip-rounded-corner-ref.html (hunk 1 above)
Comment (emilio) [reply tests/wpt/web-platform-tests/css/css-backgrounds/background-clip/clip-rounded-corner-ref.html]: Well, I guess there's no corner to hide here, so nbd, please ignore this comment.

**Comment by emilio** [comment]
> I need help to also support background images. They already have a clip from CSS as a local clip and the corner clipping should attach another clip. How is this best done?

I think @mrobinson can help you faster than I can, but I can try to look into it if he is busy :)

**Comment by mrobinson** [comment]
@pyfisch The right way to do this in WebRender world is to make the corner clip a child of the CSS clip and to ensure that display items for the border use that new clip in the clipping part of their ClipAndScrollInfo. I'm not sure how easy that is to do with our current setup, but creating clips during display list building will be important for the future in order to allow removing the rounded corner API for local clips.

**Comment by pyfisch** [comment]
Comments addressed.

The shield basically hides the rounded outer border. This is needed because the background usually covers the whole border-box and ends at the same line as the border. Due to anti aliasing the background shines through at this line and causes slightly different output for both the test and the reference. 

In current servo some white is visible between border and background. After this patch there is no more white.

test
ref

File: components/layout/display_list_builder.rs [hunk 2]
```diff
@@ -1041,6 +1030,24 @@ fn calculate_inner_bounds(mut bounds: Rect<Au>, offsets: SideOffsets2D<Au>) -> R
     bounds
 }
 
+fn calculate_inner_border_radii(
+    mut radii: BorderRadii<Au>,
+    offsets: SideOffsets2D<Au>
+    ) -> BorderRadii<Au> {
```
Comment (CYBAI) [inline components/layout/display_list_builder.rs]: Minor nit:

I think the indentation @emilio means is

```rust
fn calculate_inner_border_radii(
    mut radii: BorderRadii<Au>,
    offsets: SideOffsets2D<Au>
) -> BorderRadii<Au> {
```

Comment (emilio) [reply components/layout/display_list_builder.rs]: Yeah, that's right, that's the indentation rustfmt would use for multi-line functions.

**Comment by pyfisch** [comment]
Changed indentation.

**Review by emilio — APPROVED** [review]

**Pull request merged**